feat(treesitter)!: use extmarks to track injection regions

ribru17 opened the pull request:
**Problem:** Every time `LanguageTree:parse()` is called with a range, previously parsed injection trees are dropped, and replaced with trees parsed from the new range. Besides being inefficient and preventing injections from taking advantage of incremental reparsing, this also means that when 2+ windows are open into the same buffer, the highlighter will one at a time, meaning the trees of the others get dropped, leading to highlight flickering.

Additionally, when searching for trees to parse or trees to check for validity in `is_valid()`, currently *all* known injection regions are checked, which is inefficient for combined injections with lots of trees.

**Solution:** Track injection ranges with extmarks. This means each region is stored in a sorted BTree, so we can only iterate over the ones we need to access in a given range. This also means that we can keep trees from other ranges, since adding injection trees won't interfere with other tracked marks.

Finally fixes #32660 (for real this time :crossed_fingers:). Related: #22426. Also fixes #35993

---

Summary of changes so far:

- Only edit trees occurring on or after an edit (to adjust the included ranges)
- Use extmarks to track valid regions/trees
- Store either a `Range6` or an entire `TSTree` for a given region (depending on if the region has already been parsed), not both
- Only parse child trees when `range` is truthy; clean up the docs for `LanguageTree:parse()` since some of it is overly confusing/no longer relevant to the current implementation
- Track parsed injection regions using extmarks for better perf, correctness

TODO:

- Incrementally reparse/invalidate trees
- Maybe allow `for_each_tree` to take a `range` parameter?

Putting this up just to track the work so far. There is still much to be done. Feel free to critique as I go along, if you like

ribru17 committed:
feat(treesitter)!: use extmarks to track injection regions

Problem: Every time LanguageTree:parse() is called with a range,
previously parsed injection trees are dropped, and replaced with trees
parsed from the new range. Besides being inefficient and preventing
injections from taking advantage of incremental reparsing, this also
means that when 2+ windows are open into the same buffer, the
highlighter will one at a time, meaning the trees of the others get
dropped, leading to highlight flickering.

Additionally, when searching for trees to parse or trees to check for
validity in is_valid(), currently all known injection regions are
checked, which is inefficient for combined injections with lots of
trees.

Solution: Track injection ranges with extmarks. This means each region
is stored in a sorted BTree, so we can only iterate over the ones we
need to access in a given range. This also means that we can keep trees
from other ranges, since adding injection trees won't interfere with
other tracked marks.

Summary of changes so far:

- Only edit trees occurring on or after an edit (to adjust the included
  ranges)
- Use extmarks to track valid regions/trees
- Store either a blank `Range6` or an entire `TSTree` for a given region
  (depending on if the region has already been parsed), not both
- Only parse child trees when `range` is truthy

TODO:

- Set a flag to denote if a region is the root region
  - Tree-sitter unfortunately requires some special logic here
- Incrementally reparse/invalidate trees

justinmk commented:
A "Problem" statement will help clarify the goal here. 

p.s. does this close an open issue?

ribru17 commented:
Edited :+1: (feel free to adjust if the wording is awkward)

ribru17 commented:
This PR is now basically equivalent to the existing code; parser and highlighter tests are passing. There is just some weirdness left around conceal lines; they are trying to parse a range at `row == -1` which leads to invalid extmark placements, hence some test failures when TS is enabled for hover windows or popup menu. Looking into it now.

What's left still:

- A logic change to smartly, incrementally invalidate/reparse injected trees. Should be very simple now, hopefully a small change
- Add more info in docs since this is an invasive change
- Tests
- Keep `LanguageTree:set_included_regions()`? (maybe with a stricter API, only accepting `Range6[]` rather than lists of nodes/range4)

ribru17 committed:
fix(extmarks): always scan tree in decor_conceal_line()

**Problem:** `decor_conceal_line()` assumed that all marks placed in the
callback were conceal lines marks, meaning it was sometimes overzealous
in marking lines as concealed.

**Solution:** Always scan the tree to check for individual marks with
the `conceal_lines` attribute, rather than assuming new marks have it.

Co-authored-by: Luuk van Baal <luukvbaal@gmail.com>

ribru17 commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1774,6 +1774,9 @@ function M.open_floating_preview(contents, syntax, opts)
       local win_height = api.nvim_win_get_height(floating_winnr)
+      -- Synchronously parse the viewport in case highlighting hasn't completed. This will cause
+      -- highlighting to finish, applying conceal lines marks.
+      assert(vim.treesitter.get_parser(floating_bufnr, 'markdown')):parse({ 0, win_height })
```
You're right, sorry I must've been making a stupid mistake. Did some logging and it looks like in that final test case, the conceal line handler isn't being called with `row == 0`, even though it looks like it should since we give a `gg`. The flakiness comes from the fact that sometimes the whole viewport is parsed by `highlighter.lua` before we need to apply conceal marks. But if that doesn't happen and we have to continue parsing async, then the test fails because we don't conceal the very first line only.

I've managed to get the failure consistently on my machine by changing `30P` to `2000P`, which makes parsing go async every time (since the buffer is so large)
